### PR TITLE
Fixing and refactoring the treatment of offset rhyming patterns

### PR DESCRIPTION
**Pull request by versae**

### CODE REVIEW

File: tests/test_rhymes.py [hunk 1]
```diff
@@ -403,19 +416,19 @@ def test_get_rhymes_relaxation_offset_unrhymed(stressed_endings):
...
+        [0, -1, -2, -1, -2, -1, -2, -1, -2, -1, -2, -1, -2, -1, 0, -1]
     )
     assert get_rhymes(
         stressed_endings, assonance=True, relaxation=True, offset=4,
         unrhymed_verse_symbol="$"
     ) == output
 
 
-def test_get_rhymes_assonance_relaxation_exceded_offset_unrhymed():
+def test_get_rhymes_assonance_relaxation_exexceededffset_unrhymed():
```
Comment (versae) [inline tests/test_rhymes.py]:
```suggestion
def test_get_rhymes_assonance_relaxation_exceeded_offset_unrhymed():
```

File: src/rantanplan/utils.py [hunk 2]
```diff
@@ -0,0 +1,24 @@
+# -*- coding: utf-8 -*-
+from collections import Counter
+
+
+def generate_exceeded_offset_indices(values, offset=4):
+    """Yield the indices of the duplicated elements
+    at offset + 1 positions from their previous one; or -1 if none found.
```
Comment (versae) [inline src/rantanplan/utils.py]:
```suggestion
    at offset + 1 positions from their previous one; or None if none found.
```

File: src/rantanplan/rhymes.py [hunk 3]
```diff
@@ -121,47 +122,45 @@ def get_clean_codes(stressed_endings, assonance=False, relaxation=False):
...
-    lines 1 and 100 if the ending is the same. To control how many lines
-    should a matching rhyme occur in, an offset can be set to an arbitrary
-    number, effectively allowing rhymes that only occur between
-    lines i and i + offset."""
-    letters = {}
+def apply_offset(codes, ending_codes, offset=4):
+    """Control how many lines of distnace should a matching rhyme occur at.
+    An offset can be set to an arbitrary number, effectively allowing rhymes
+    that only occur between lines i and i + offset, and assigning a new rhyme
+    code when the offser is exceeded, even if the ending appeared before."""
```
Comment (versae) [inline src/rantanplan/rhymes.py]:
```suggestion
    """Control how many lines of distance should a matching rhyme occur at.
    An offset can be set to an arbitrary number, effectively allowing rhymes
    that only occur between lines i and i + offset, and assigning a new rhyme
    code when the offset is exceeded, even if the ending appeared before."""
```

**Review by alvp — APPROVED** [review]
Perfect, as always ;)

**Pull request merged**